nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetChangeDataCaptureMySQL.java

```diff
@@ -0,0 +1,879 @@
+                            throw new IOException("BEGIN event received while already processing a transaction. This could indicate that your binlog position is invalid.");
+                        }
+                        // Mark the current binlog position in case we have to rollback the transaction (if the processor is stopped, e.g.)
+                        xactBinlogFile = currentBinlogFile;
+                        xactBinlogPosition = currentBinlogPosition;
+                        xactSequenceId = currentSequenceId.get();
+
+                        BeginTransactionEventInfo beginEvent = new BeginTransactionEventInfo(timestamp, currentBinlogFile, currentBinlogPosition);
+                        currentSequenceId.set(beginEventWriter.writeEvent(currentSession, beginEvent, currentSequenceId.get()));
+                        inTransaction = true;
+                    } else if ("COMMIT".equals(sql)) {
+                        if (!inTransaction) {
+                            throw new IOException("COMMIT event received while not processing a transaction (i.e. no corresponding BEGIN event). "
+                                    + "This could indicate that your binlog position is invalid.");
+                        }
+                        // InnoDB generates XID events for "commit", but MyISAM generates Query events with "COMMIT", so handle that here
+                        CommitTransactionEventInfo commitTransactionEvent = new CommitTransactionEventInfo(timestamp, currentBinlogFile, currentBinlogPosition);
+                        currentSequenceId.set(commitEventWriter.writeEvent(currentSession, commitTransactionEvent, currentSequenceId.get()));
+                        // Commit the NiFi session
+                        session.commit();
+                        inTransaction = false;
+                        currentTable = null;
+
+                    } else {
+                        // Check for schema change events (alter table, e.g.). Normalize the query to do string matching on the type of change
+                        String normalizedQuery = sql.toLowerCase().trim().replaceAll(" {2,}", " ");
+                        if (normalizedQuery.startsWith("alter table")
+                                || normalizedQuery.startsWith("alter ignore table")
+                                || normalizedQuery.startsWith("drop table")) {
+                            SchemaChangeEventInfo schemaChangeEvent = new SchemaChangeEventInfo(currentTable, timestamp, currentBinlogFile, currentBinlogPosition, normalizedQuery);
+                            currentSequenceId.set(schemaChangeEventWriter.writeEvent(currentSession, schemaChangeEvent, currentSequenceId.get()));
+                            // Remove all the keys from the cache that this processor added
+                            if (cacheClient != null) {
+                                cacheClient.removeByPattern(this.getIdentifier() + ".*");
```

Comment:
This is an unfortunate consequence of the fact that unless we parse the SQL statement, we don't know which table is changing (and how), so we don't know which of our entries to update/refresh. So instead we wipe the whole list and build up the cache again.